### ISSUE TITLE
[management] Role mapping issue for Oauth2 authentication

### ISSUE THREAD

**Issue by vlemaitre** [gravitee-io/issues#1727]
I am not able to map roles to user authenticated through keycloak identity provider.
Here is an extract of the user JWT token : 
```
{
  "resource_access": {
    "gravitee": {
      "roles": [
        "ADMIN"
      ]
    },
    "account": {
      "roles": [
        "manage-account",
        "manage-account-links",
        "view-profile"
      ]
    }
  },
  "scope": "email profile",
  "email_verified": false,
  "job_id": "GRAVITEE_ADMIN",
  "name": "toto TOTO",
  "preferred_username": "toto",
  "given_name": "toto",
  "family_name": "TOTO",
  "email": "toto@toto.fr"
}
```

I try to configure my oauth2 api security provider using the role mapping features as implemented here : 
[pull request 469](https://github.com/gravitee-io/gravitee-management-rest-api/pull/469)

Extract of my gravitee.yml configuration file : 
```
mapping:
  id: sub
  email: email
  lastname: family_name
  firstname: given_name
  roles:
  - mapping:
      condition: "{#jsonPath(#profile, '$.job_id') == 'GRAVITEE_ADMIN'}"
      values: ["portal:admin", "management:admin"]
```

### Expected Behavior
The user named 'toto' should be created with the roles "PORTAL:ADMIN" and "MANAGEMENT:ADMIN" because its 'job_id' attribute is equal to "GRAVITEE_ADMIN".

### Current Behavior
The 'toto' user is created with the default roles "PORTAL:USER" and "MANAGEMENT:USER".

The issue is the same when I tried to map the roles from the attribute "resource_access.gravitee.roles".

### Possible Solution
Maybe my mappings definition are wrong.
Or there is a bug.
Please explain me how to configure this role mappings.

### Steps to Reproduce (for bugs)
1. Configure your ouath2 identity provider to add specific attributes or roles to a user
2. Configure your oauth2 identity provider in order the token generated for gravitee client contains these user attributes or roles
3. Configure the role mappings on oauth2 security provider into Gravitee API configuration.
4. Login using the user with the matching attributes or roles.

### Context
It obliges to redefine the user permissions into Gravitee applications.
We would like to centralize the user roles into our Identity Provider solution which is Keycloak.

### Your Environment
* Version used: 1.20.0
* Browser Name and version: Chrome latest
* Operating System and version: Docker images running on Kubernetes

**Comment by vlemaitre** [comment]
When I enable the debug traces on AuthorizationServerConfigurationParser, I am able to see that the mapping seems to be correct : 
[gravitee-listener-36] DEBUG i.g.m.r.r.a.o.AuthorizationServerConfigurationParser - Expression {#jsonPath(#profile, '$.job_id') == 'GRAVITEE_ADMIN'} give roles [portal:admin, management:admin]

But these roles are finally not assigned to my user.

**Comment by brasseld** [comment]
Hi Vincent,

The log you're providing is the one print during parsing time (I mean parsing of the configuration) and does not reflect what is done during user authentication.

**Comment by vlemaitre** [comment]
Sorry you are right this trace occurs only 1 time at the first authentication.

**Comment by vlemaitre** [comment]
Hi,

With version 1.21.2, the role mapping works correctly.
But I am  not sure to understand how I have to configure an identity provider (keycloak for instance) on this version...
I see that "oauth2" type providers can now be added from admin settings menu of manager UI.

Is it still possible to define and configure an oauth2 identity provider as explained into your [documentation](https://docs.gravitee.io/apim_installguide_authentication_keycloak.html) ?
I do not succeed to configure Keycloak by this way, since I updated to 1.21.x release.

**Comment by brasseld** [comment]
Hi @vlemaitre 

Here are some screenshots to explain you how to configure Keycloak access:

Create Realm
<img width="1170" alt="capture d ecran 2018-12-10 a 14 03 18" src="https://user-images.githubusercontent.com/133094/49734700-70e97c80-fc85-11e8-8057-c358a848f494.png">

Create client
<img width="1131" alt="capture d ecran 2018-12-10 a 14 03 47" src="https://user-images.githubusercontent.com/133094/49734709-7c3ca800-fc85-11e8-997e-2e661bfee57d.png">

Create user
<img width="1190" alt="capture d ecran 2018-12-10 a 14 10 20" src="https://user-images.githubusercontent.com/133094/49734721-85c61000-fc85-11e8-9322-a2c1e2188b9d.png">

Then, configure identity provider (type OIDC) in Gravitee:

<img width="1021" alt="capture d ecran 2018-12-10 a 14 12 15" src="https://user-images.githubusercontent.com/133094/49734761-b017cd80-fc85-11e8-9dc9-bf7857590645.png">
<img width="1009" alt="capture d ecran 2018-12-10 a 14 12 22" src="https://user-images.githubusercontent.com/133094/49734768-b1e19100-fc85-11e8-9ec2-4090ef3adb8c.png">
<img width="1030" alt="capture d ecran 2018-12-10 a 14 12 28" src="https://user-images.githubusercontent.com/133094/49734769-b443eb00-fc85-11e8-8390-5c9f411932ee.png">

Then logout from admin account. From the login view you must have a new button for Keycloak
<img width="663" alt="capture d ecran 2018-12-10 a 14 10 32" src="https://user-images.githubusercontent.com/133094/49734797-c6258e00-fc85-11e8-9fc1-f8a25568864b.png">

Login through Keycloak, and you're automatically redirected to gravitee
<img width="258" alt="capture d ecran 2018-12-10 a 14 10 38" src="https://user-images.githubusercontent.com/133094/49734830-d63d6d80-fc85-11e8-8722-fa07c79da04f.png">

**Comment by vlemaitre** [comment]
I do the configuration like this. And it works.

But I amwondering if it is also possible to configure it directly from the config.yml file as explained [here](https://docs.gravitee.io/apim_installguide_authentication_keycloak.html).
This configuration was working in release 1.20.0, but not with release 1.21.x.

To sum up:
- in release 1.20.0 : 
         - keycloak configuration into config.yml : OK
         - role mapping : KO
- in release 1.21.x : 
         - keycloak configuration into config.yml : KO
         - role mapping : OK

Thanks

**Comment by vlemaitre** [comment]
We are using Helm / Kubernetes to deploy Gravitee.
We would like to get all the configuration into a file (config.yml) which is included into a Kubernetes config map.
By this way, the deployment of the Gravitee application can be fully-configured and managed at Kubernetes level. Ops just have to update the Keycloak endpoints to start the application, and do not need to access the Gravitee UI.

**Comment by brasseld** [comment]
Ok, in that case, initializing the platform by calling its API (ie. Management API) would not be a solution ?

**Comment by vlemaitre** [comment]
Yes I think this is the solution.
I will test it.
Thanks for your help.

**Comment by brasseld** [comment]
Do you need an example  to create an identity provider from API?

**Comment by vlemaitre** [comment]
If you have an example it will help me ;-)

**Comment by brasseld** [comment]
Just POST this kind of content:

```json
{
	"name": "Keycloak Demo",
    "type": "OIDC",
    "configuration": {
        "clientId": "portal",
        "clientSecret": "bbee50c7-e98c-4731-ab20-12c0b8e6e39a",
        "color": "#9a3a3a",
        "scopes": [
            "openid"
        ],
        "authorizeEndpoint": "http://localhost:8080/auth/realms/Gravitee/protocol/openid-connect/auth",
        "tokenEndpoint": "http://localhost:8080/auth/realms/Gravitee/protocol/openid-connect/token",
        "tokenIntrospectionEndpoint": "http://localhost:8080/auth/realms/Gravitee/protocol/openid-connect/token/introspect",
        "userInfoEndpoint": "http://localhost:8080/auth/realms/Gravitee/protocol/openid-connect/userinfo",
        "userLogoutEndpoint": "http://localhost:8080/auth/realms/Gravitee/protocol/openid-connect/logout?redirect_uri="
    },
    "enabled": true,
    "userProfileMapping": {
    	"id": "sub",
        "email": "email",
        "firstname": "given_name",
        "lastname": "family_name",
        "picture": "picture"
    }
}
```

**Comment by brasseld** [comment]
Also, if you can explain me what can be the best way to run a script within Kubernetes while creating the platform, I would really appreciate it ;)

**Comment by vlemaitre** [comment]
Thanks for the example.
For the moment I do not know how to call this API.
I think I will add a "config" container that just execute API calls to setup the Gravitee configuration. 
But it means I have to keep at least one "memory" user definition with admin right in order to be able to call this API and config the external identity provider.

So I will study how to proceed ;-) 
I will keep you informed about the method.

And what is the reason to remove the possibility to configure the identity provider from config.yml file ?
It was not compatible with the API ?

**Comment by brasseld** [comment]
First reason was that we need to support multiple authentication providers. It was not the case before 1.21

But we know that being able to configure the platform by providing such yml configuration is the trick and we have to provide such support. We are just thinking what would be the ideal way (and also, how to manage delta between a given conf and what is stored in database...)

**Comment by vlemaitre** [comment]
I understand.

Another question : when are executed the identity provider mapping rules ?
At each authentication ? Or only once, at the first user authentication ?

The usecase : 
1- I configure a user with "API_PUBLISHER" role in keycloak.
2- I configure the role mappings in Gravitee to manage API_PUBLISHER and ADMIN roles.
3- First login with this user : the role in Gravitee are ok.
4- I update the user role in Keycloak to "ADMIN"
5- Second login with this user : the role in Gravitee is not updated.

**Comment by brasseld** [comment]
The question has already been asked: https://github.com/gravitee-io/issues/issues/1698

We must work on the concept of "dynamic group" and "dynamic role" to make them updatable during each user authentication.

**Comment by vlemaitre** [comment]
I have another issue to get my role mapping working.
Based on the JWT access token delivered by Keycloak I sent at the beginning of this thread, 
conditioning the mapping on a custom attribute "job_id" works fine.
Here is my jsonPath expression: 
```
{#jsonPath(#profile, '$.job_id') == "GRAVITEE_ADMIN"}
```

I tried to write the jsonPath expression to condition on "resource_access.gravitee.roles", but without success. Each time I get an exception because property is missing in path 'resource_access'.
Here is my jsonPath expression : 
```
{#jsonPath(#profile, '$.resource_access.gravitee.roles[0]') == "ADMIN"}
```
This expression works fine on a jsonPath tester with my JWT Token.

Consequently, I am wondering if the access token received by Gravitee from Keycloak is correct.
I am not able to capture it using Chrome developper tool because the keycloak login windows close immediately.
Could you tell how I can inspect this access token ?

Thanks

**Comment by brasseld** [comment]
Be careful @vlemaitre 

We are not relying on the JWT provided by Keycloak but we are invoking the UserInfo exposed by Keycloak to get the userprofile of the authenticated user.
So you'll probably have to configure mapping for your client, to retrieve data you need to apply role / group mapping.

**Comment by vlemaitre** [comment]
Thanks a lot.
I missed this point.
It works now.

Best Regards

**Comment by brasseld** [comment]
Well done!

**Comment by n2ygk** [comment]
@vlemaitre (CC @brasseld) Can you explain how you ended up doing this? I am also trying to use fields in the userinfo response to map roles. For example, if my userinfo response is:
```json
{
    "sub": "ac45@columbia.edu",
    "https://api.columbia.edu/claim/group": "api_publisher something_else etc",
    "active": true,
    "token_type": "Bearer",
    "given_name": "Alan",
    "client_id": "gravitee",
    "uid": "ac45@columbia.edu",
    "scope": "email openid profile auth-columbia https://api.columbia.edu/scope/group",
    "name": "Alan Crosswell",
    "exp": 1554748851,
    "family_name": "Crosswell",
    "email": "alan@columbia.edu",
    "username": "ac45@columbia.edu"
}
```
Would my expression be something like this?
```
{#jsonPath(#userinfo, "$.'https://api.columbia.edu/claim/group'").contains("api_publisher")}
```

**Comment by n2ygk** [comment]
PS: Is there a document or source file somewhere that lists all the defined variables that this stuff uses? It's not `#userinfo` or `#response.content` :-(

**Comment by n2ygk** [comment]
Found [it!](https://github.com/gravitee-io/gravitee-management-rest-api/blob/6935c23011278a84183ddeb5a72bc3cc8e5384ae/gravitee-management-api-rest/src/main/java/io/gravitee/management/rest/resource/auth/OAuth2AuthenticationResource.java#L398-L432)

`userInfo` response gets mapped to `#profile`

```
{#jsonPath(#profile, '$.sub').contains("ac45")}
```

**Comment by brasseld** [comment]
Nice shot @n2ygk 

You make me think it would be better to use #userinfo instead of #profile... or both for backward compatibility....

**Comment by n2ygk** [comment]
I think mostly some documentation updates would do the trick.

BTW, I'm running into an error where my JSON key looks like this (for local namespacing): `https://api.columbia.edu/claim/group` and am getting this error:

```
Caused by: org.springframework.core.convert.ConversionFailedException: Failed to convert from type [java.lang.String] to type [boolean] for value '{jsonPath(#profile, "$['http://api.columbia.edu/claim/group']") ? #jsonPath(#profile, "$['http://api.columbia.edu/claim/group']").contains("test_test1") : false}'; nested exception is java.lang.IllegalArgumentException: Invalid boolean value '{jsonPath(#profile, "$['http://api.columbia.edu/claim/group']") ? #jsonPath(#profile, "$['http://api.columbia.edu/claim/group']").contains("test_test1") : false}'
	at org.springframework.core.convert.support.ConversionUtils.invokeConverter(ConversionUtils.java:46)
	at org.springframework.core.convert.support.GenericConversionService.convert(GenericConversionService.java:191)
	at org.springframework.expression.spel.support.StandardTypeConverter.convertValue(StandardTypeConverter.java:70)
	... 98 common frames omitted
Caused by: java.lang.IllegalArgumentException: Invalid boolean value '{jsonPath(#profile, "$['http://api.columbia.edu/claim/group']") ? #jsonPath(#profile, "$['http://api.columbia.edu/claim/group']").contains("test_test1") : false}'
	at org.springframework.core.convert.support.StringToBooleanConverter.convert(StringToBooleanConverter.java:63)
	at org.springframework.core.convert.support.StringToBooleanConverter.convert(StringToBooleanConverter.java:31)
	at org.springframework.core.convert.support.GenericConversionService$ConverterAdapter.convert(GenericConversionService.java:385)
	at org.springframework.core.convert.support.ConversionUtils.invokeConverter(ConversionUtils.java:40)
```

This is what I tried to use: 
```
{jsonPath(#profile, "$['http://api.columbia.edu/claim/group']") ? #jsonPath(#profile, "$['http://api.columbia.edu/claim/group']").contains("test_test1") : false}
```
Not sure if this stuff supports the Elvis operator....

**Comment by n2ygk** [comment]
D'oh never mind. I had the key wrong -- it's `https:` not `http:`. Much simpler seems to work: `{#jsonPath(#profile, "$['https://api.columbia.edu/claim/group']").contains('test_test1')}`

Still wondering how to detect if the key is missing from the userinfo response, but I believe it will always be there as long as I request the matching scope.

**Comment by brasseld** [comment]
As fart as I know, EL is supporting Elvis operator.

The problem here is that the property's key is a URI.

When applying for selection using jsonPath, you must do something like this:

```
{#jsonPath(#profile, "$.'https://api.columbia.edu/claim/group'")
```

**Comment by n2ygk** [comment]
per https://github.com/json-path/JsonPath/issues/482 that doesn't work but the brackets notation does.  One would expect that form to work, but it doesn't.

**Comment by brasseld** [comment]
Perfect, thanks for the tip!

**Comment by n2ygk** [comment]
@brasseld BTW, where do the mappings get persisted. Mongo has this for `identity_providers`:
```json
{
...
  "roleMappings": {
    "eyBzdWIgPT0gJ2FjNDVAY29sdW1iaWEuZWR1JyB9": [
      "2:ADMIN",
      "1:ADMIN"
    ]
  },
...
}
```

I'm trying to make a backup of my config before I break it with my next experiment:-)

**Comment by brasseld** [comment]
The condition (`eyBzdWIgPT0gJ2FjNDVAY29sdW1iaWEuZWR1JyB9`) is base64 encoded to avoid issue with Mongo syntax, as we are using the condition as the key.

**Comment by n2ygk** [comment]
Ah, that was easy! Thanks.

**Comment by urs86** [comment]
I have created 2 oauth2 plans. 
One for admin and one for regular user. 
I've tried to apply a different rule for my plans in order to give path access only to "GET" methods for a regular user and "GET", "POST" for an admin user. 
Just for test purposes. I've tried with both ROLE / USER NAME, but it seems not working. 
When applying the 'ROLE' as a filter, nothing is accessible. That's 'ok'... perhaps(sure) my EL rule is wrong...
#context.attributes['jwt'].claims['resource_access.my-api.roles'] == 'admin'
#context.attributes['jwt'].claims['resource_access']['my-api']['roles'] == 'admin'
#context.attributes['jwt'].claims['resource_access']['my-api']['roles'][0] == 'admin'

I tried with different USER NAM as a rule and EVERYTHING is accessible for BOTH regular and admin users... That's not ok at all. My EL is correct 
#context.attributes['jwt'].claims['preferred_username'] == 'admin_1' -> applied to my oauth2 admin plan
#context.attributes['jwt'].claims['preferred_username'] == 'user_1' -> applied to my oauth2 user plan

Je ne sais pas ce qu'il faut faire...

**Comment by aelamrani** [comment]
Hi @urs86 

I'd say this one should work if `admin` is the first element: `#context.attributes['jwt'].claims['resource_access']['my-api']['roles'][0]`

Can you provide an example of claim `resource_access`?

**Comment by urs86** [comment]
My bad @aelamrani  for the above post. Indeed, in resource_access I don't have an admin role...
Indeed, I've tried with #context.attributes['jwt'].claims['realm_access']['roles'][0] == 'admin' and it works.
`"realm_access": {
    "roles": [
      "admin",
      "user"
    ]
  },
  "resource_access": {
    "account": {
      "roles": [
        "manage-account",
        "manage-account-links",
        "view-profile"
      ]
    }
  },
 ...
"preferred_username": "admin_1"
`
Though, within the OAUTH2 context, if I need to allow some users with ROLE_1 (let's say) to API_1 (throughout Path Authorization) to access only GET methods and another users with ROLE_2 to same API_1 to access GET/POST methods, there is no way now. 

I can't subscribe to the second plan : "An other OAuth2 or JWT plan is already subscribed by the same application". 

If I create a new APP, but with the same keycloak client ID (my-api) : "An application already exists with client_id [my-api]"

Perhaps I don't understand exactly the way it was designed, but I would not create a new client in keycloak, a new API and APP in gravitee management, in order to filter the access for some users to some different METHODS of the same endpoint finally (or paths). I'll continue the research on how to limit the access based on roles.

Though, a big thanks for your answer @aelamrani and you've done such a good job until now guys!
Best!

**Comment by aelamrani** [comment]
Why not use OAuth2 scopes? And use the [role base access policy](https://github.com/gravitee-io/gravitee-policy-role-based-access-control)?

**Comment by urs86** [comment]
Thanks a lot @aelamrani ! I'll try this one.

**Comment by aelamrani** [comment]
You are welcome!
